File: 315.Count-of-Smaller-Numbers-After-Self.java
// https://leetcode.com/problems/count-of-smaller-numbers-after-self/
//
// algorithms
// Hard (37.96%)
// Total Accepted:    74,842
// Total Submissions: 197,164
// beats 92.20% of java submissions
// 这道题开始一直没有用 mapping 数值来保存下标，res 一直加错索引，后面断点调试发现了

class Solution {
    public List<Integer> countSmaller(int[] nums) {
        int len = nums.length;
        ArrayList<Integer> l = new ArrayList<>(len);
        if (len == 0) {
            return l;
        }
        int[] res = new int[len];
        int[] mapping = new int[len];

        for (int i = 0; i < len; i++) {
            mapping[i] = i;
        }

        mergeSort(nums, mapping, res, 0, nums.length - 1);

        for (int i = 0; i < len; i++) {
            l.add(res[i]);
        }

        return l;
    }

    public void mergeSort(int[] nums, int[] mapping, int[] res, int head, int tail) {
        if (tail == head) {
            return;
        }

        int middle = head + (tail - head) / 2;
        mergeSort(nums, mapping, res, head, middle);
        mergeSort(nums, mapping, res,middle + 1, tail);

        int l = head, r = middle + 1;
        int[] tmp = new int[tail - head + 1];
        int idx = 0;
        while (l <= middle && r <= tail) {
            if (nums[mapping[l]] > nums[mapping[r]]) {
                tmp[idx] = mapping[l];
                res[mapping[l]] += (tail - r + 1);
                l++;
            } else {
                tmp[idx] = mapping[r];
                r++;
            }
            idx++;
        }

        while (l <= middle) {
            tmp[idx++] = mapping[l++];
        }
        while (r <= tail) {
            tmp[idx++] = mapping[r++];
        }

        for (int i = head; i <= tail; i++) {
            mapping[i] = tmp[i - head];
        }

    }

}

File: 315.Count-of-Smaller-Numbers-After-Self.py
# https://leetcode.com/problems/count-of-smaller-numbers-after-self/
#
# algorithms
# Hard (39.35%)
# Total Accepted:    88,802
# Total Submissions: 225,695


class Solution(object):
    def countSmaller(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        length = len(nums)
        if length == 0:
            return []

        origin_idx = [i for i in xrange(length)]  # 存储当前这个位置元素在原始数组中的位置
        more = [0] * length

        def merge_sort(start, end):
            if start >= end:
                return

            middle = start + (end - start) / 2
            merge_sort(start, middle)
            merge_sort(middle + 1, end)

            tmp = []
            origin_tmp = []
            l_idx, r_idx = start, middle + 1
            while l_idx <= middle and r_idx <= end:
                if nums[l_idx] > nums[r_idx]:
                    tmp.append(nums[l_idx])
                    more[origin_idx[l_idx]] += (end - r_idx + 1)
                    origin_tmp.append(origin_idx[l_idx])
                    l_idx += 1
                else:
                    tmp.append(nums[r_idx])
                    origin_tmp.append(origin_idx[r_idx])
                    r_idx += 1

            if l_idx <= middle:
                tmp += nums[l_idx:middle + 1]
                origin_tmp += origin_idx[l_idx:middle + 1]
            if r_idx <= end:
                tmp += nums[r_idx:end + 1]
                origin_tmp += origin_idx[r_idx:end + 1]

            for i in xrange(start, end + 1):
                nums[i] = tmp[i - start]
                origin_idx[i] = origin_tmp[i - start]

        merge_sort(0, length - 1)

        return more
